theano/gpuarray/dnn.py

```diff
@@ -2732,6 +2732,189 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+    """
+    This Op builds a spatial transformer descriptor for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ('dimensions', 'precision')
+    params_type = ParamsType(nimages=int_t, height=int_t, width=int_t, nchannels=int_t,
+                             nb_dims=int_t, precision=cudnn.cudnnDataType_t)
+
+    def c_headers(self):
+        return ['cudnn.h', 'cudnn_helper.h']
+
+    def c_header_dirs(self):
+        return [os.path.dirname(__file__), config.dnn.include_path]
+
+    def c_libraries(self):
+        return ['cudnn']
+
+    def c_lib_dirs(self):
+        return [config.dnn.library_path]
+
+    def do_constant_folding(self, node):
+        return False
+
+    def __init__(self, dimensions, precision="float32"):
+        COp.__init__(self, ["c_code/spatialtf_desc.c"], "APPLY_SPECIFIC(spatialtf_desc)")
+
+        # dimensions must have at least width and height
+        assert len(dimensions) >= 2
+
+        self.dimensions = tuple(dimensions)
+
+        # cuDNN supports only 2D transformations, therefore output tensor must
+        # not exceed 4 dimensions (width, height, num_feature_maps, num_images)
+        assert len(self.dimensions) <= 4
+
+        assert cudnn.cudnnDataType_t.has_alias(precision)
+        self.precision = precision
+
+    def make_node(self):
+        node = Apply(self, [],
+                     [CDataType("cudnnSpatialTransformerDescriptor_t",
+                                freefunc="cudnnDestroySpatialTransformerDescriptor")()])
+        # DebugMode cannot compare the values of CDataType variables, so by
+        # default it returns False all the time. To prevent DebugMode from
+        # complaining because of the MergeOptimizer, we make this variable
+        # always compare to True.
+        out = node.outputs[0]
+        out.tag.values_eq_approx = tensor.type.values_eq_approx_always_true
+        return node
+
+    # Grid width
+    nimages = property(lambda self: self.dimensions[0])
+    # Grid height
+    height = property(lambda self: self.dimensions[1])
+    # Number of feature maps
+    width = property(lambda self: self.dimensions[2])
+    # Number of images
+    nchannels = property(lambda self: self.dimensions[3])
+    # Number of dimensions in the output tensor
+    nb_dims = property(lambda self: len(self.dimensions))
+
+    def c_code_cache_version(self):
+        return (super(GpuDnnSpatialTfDesc, self).c_code_cache_version(), version())
+
+
+class GpuDnnGridGenerator(DnnBase):
+
+    """
+    This Op builds a spatial transformer grid generator for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ()
+    _cop_num_inputs = 3
+    _cop_num_outputs = 1
+
+    def __init__(self):
+        DnnBase.__init__(self, ["c_code/spatialtf_grid.c"], "spatialtf_grid")
+
+    def dnn_context(self, node):
+        return node.outputs[0].type.context_name
+
+    def make_node(self, grid_dimensions, theta, desc, precision, cx=None):
```

Comment:
I've implemented these suggestions, and fixed the parameter ordering in the C op as well.

Yes, `precision` should be the same cuDnnDataType_t as specified in the spatial transformer descriptor.